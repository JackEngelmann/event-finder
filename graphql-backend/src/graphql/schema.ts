import { gql } from 'apollo-server-express'

export const typeDefs = gql`
    type Club {
        id: Int!
        name: String!
    }

    type Event {
        id: Int!
        name: String!
        description: String
        date: String!
        club: Club!
        imageUrl: String
    }

    type Query {
        club(id: Int!): Club
        clubs: [Club!]!
        event(id: Int!): Event
        events: [Event!]!
    }
`
