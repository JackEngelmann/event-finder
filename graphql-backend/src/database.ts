import sqlite3 from 'sqlite3'
import { insertClubsIfNoneExist } from './seed-data/clubData'
import { insertEventsIfNoneExist } from './seed-data/eventData'

sqlite3.verbose()

const DB_SOURCE = 'db.sqlite'
export const db = new sqlite3.Database(DB_SOURCE, async err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    db.serialize(() => {
        createTables()
        insertClubsIfNoneExist(db)
        insertEventsIfNoneExist(db)
    })
})

async function createTables() {
    db.run(`
        CREATE TABLE IF NOT EXISTS club (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text
        )
    `)
    db.run(`
        CREATE TABLE IF NOT EXISTS event (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text,
            description text,
            date string,
            clubId integer,
            imageUrl text
        )
    `)
}
