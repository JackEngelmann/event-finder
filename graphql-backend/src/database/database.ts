import sqlite3, { Database } from 'sqlite3'
import { createSchema } from './createSchema'
import { applySeeds } from './applySeeds'

sqlite3.verbose()

const DB_SOURCE = 'db.sqlite'

export const db = new sqlite3.Database(DB_SOURCE, async err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    db.serialize(() => {
        createSchema(db)
        applySeeds(db)
    })
})

export const createTestDb = async() => new Promise<Database>(resolve => {
    const db = new sqlite3.Database(':memory:', async err => {
        if (err) {
            console.error(err.message)
            throw err
        }
    })
    db.serialize(async () => {
        createSchema(db)
        resolve(db)
    })
})
