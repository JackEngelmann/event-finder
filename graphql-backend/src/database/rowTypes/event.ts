export type EventRow = {
    id: number
    name: string
    description: string
    clubId: string
    date: string
    priceCategory: 1 | 2 | 3
    admissionFee: number
    admissionFeeWithDiscount: number
    special: string
    minimumAge: number
    amountOfFloors: number
}