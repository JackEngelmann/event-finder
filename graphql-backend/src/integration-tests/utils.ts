import { Database } from "sqlite3"

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name,
    description,
    date,
    imageUrl,
    priceCategory,
    admissionFee,
    admissionFeeWithDiscount,
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    name
    address
    region
    contact
    email
    specials
    description
    link
`

export function insertTestData(db: Database) {
    return new Promise(resolve => {
        db.serialize(() => {
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO event
                    (
                        name,
                        description,
                        date,
                        clubId,
                        imageUrl,
                        priceCategory,
                        admissionFee,
                        admissionFeeWithDiscount,
                        special,
                        minimumAge,
                        amountOfFloors
                    ) 
                    VALUES (
                        'name',
                        'description',
                        'date',
                        1,
                        'imageUrl',
                        2,
                        10.0,
                        20.0,
                        'specials',
                        18,
                        4
                    )
                `)
            db.run(`
                    INSERT INTO genre ([name]) VALUES
                        ('genre1'),
                        ('genre2'),
                        ('genre3')
                `)
            db.run(
                `
                    INSERT INTO eventGenre (eventId, genreId) VALUES (1, 1)
                `,
                () => resolve()
            )
        })
    })
}
