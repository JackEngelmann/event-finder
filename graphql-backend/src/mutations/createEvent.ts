import { AppContext } from '../appContext'
import { Database } from 'sqlite3'

export type CreateEventInput = {
    name: string
    description?: String
    date: string
    clubId: number
    genreIds?: number[]
    special?: string
    priceCategory?: number
    admissionFee?: number
    admissionFeeWithDiscount?: number
    minimumAge?: number
    amountOfFloors?: number
}

// TODO: use $ for db params

export function createEvent(appContext: AppContext, input: CreateEventInput) {
    const { db } = appContext
    return new Promise<number>((resolve, reject) => {
        db.serialize(async () => {
            const eventId = await insertEventRow(db, input)
            await assignGenres(db, eventId, input.genreIds)
            resolve(eventId)
        })
    })
}

function insertEventRow(db: Database, input: CreateEventInput) {
    return new Promise<number>((resolve, reject) => {
        db.run(
            `
                INSERT INTO event (
                    name,
                    date,
                    description,
                    clubId,
                    special,
                    priceCategory,
                    admissionFee,
                    admissionFeeWithDiscount,
                    minimumAge,
                    amountOfFloors
                ) VALUES (
                    $name,
                    $date,
                    $description,
                    $clubId,
                    $special,
                    $priceCategory,
                    $admissionFee,
                    $admissionFeeWithDiscount,
                    $minimumAge,
                    $amountOfFloors
                )
            `,
            {
                $name: input.name,
                $date: input.date,
                $description: input.description,
                $clubId: input.clubId,
                $special: input.special,
                $priceCategory: input.priceCategory,
                $admissionFee: input.admissionFee,
                $admissionFeeWithDiscount: input.admissionFeeWithDiscount,
                $minimumAge: input.minimumAge,
                $amountOfFloors: input.amountOfFloors,
            },
            function(err) {
                if (err) reject(err)
                resolve(this.lastID)
            }
        )
    })
}

function assignGenres(
    db: Database,
    eventId: number,
    genreIds: number[] | undefined
) {
    return new Promise((resolve, reject) => {
        if (genreIds === undefined) return resolve()
        const placeholders = genreIds.map(g => '(?, ?)').join(',')
        const values = genreIds.flatMap(genreId => [eventId, genreId])
        db.run(
            'insert into eventGenre (eventId, genreId) VALUES ' + placeholders,
            values,
            err => err ? reject(err) : resolve(),
        )
    })
}
