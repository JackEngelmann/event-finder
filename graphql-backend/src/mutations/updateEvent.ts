import { AppContext } from '../appContext'
import { Database } from 'sqlite3'
import { requireNumber } from '../validationUtils'

export type UpdateEventInput = {
    id: number
    name: string
    description?: String
    date: string
    clubId: number
    genreIds?: number[]
    special?: string
    priceCategory?: number
    admissionFee?: number
    admissionFeeWithDiscount?: number
    minimumAge?: number
    amountOfFloors?: number
}

export function updateEvent(appContext: AppContext, input: UpdateEventInput) {
    const { db } = appContext
    return new Promise((resolve, reject) => {
        db.serialize(async () => {
            await updateEventRow(db, input)
            await updateGenreAssignments(db, input.id, input.genreIds || [])
            resolve()
        })
    })
}

function updateEventRow(db: Database, input: UpdateEventInput) {
    requireNumber(input.id)
    return new Promise<number>((resolve, reject) => {
        db.run(
            `
                UPDATE event 
                SET
                    admissionFee = $admissionFee,
                    admissionFeeWithDiscount = $admissionFeeWithDiscount,
                    amountOfFloors = $amountOfFloors,
                    clubId = $clubId,
                    date = $date,
                    description = $description,
                    minimumAge = $minimumAge,
                    name = $name,
                    priceCategory = $priceCategory,
                    special = $special
                WHERE id = $id
            `,
            {
                $admissionFee: input.admissionFee,
                $admissionFeeWithDiscount: input.admissionFeeWithDiscount,
                $amountOfFloors: input.amountOfFloors,
                $clubId: input.clubId,
                $date: input.date,
                $description: input.description,
                $id: input.id,
                $minimumAge: input.minimumAge,
                $name: input.name,
                $priceCategory: input.priceCategory,
                $special: input.special,
            },
            function(err) {
                if (err) reject(err)
                resolve(this.lastID)
            }
        )
    })
}

function updateGenreAssignments(
    db: Database,
    eventId: number,
    genreIds: number[] 
) {
    return new Promise((resolve, reject) => {
        const placeholders = genreIds.map(g => '(?, ?)').join(',')
        const values = genreIds.flatMap(genreId => [eventId, genreId])
        db.serialize(() => {
            db.run(
                `DELETE FROM eventGenre WHERE eventId = $eventId`,
                {$eventId: eventId}
            )
            if (genreIds.length === 0) return resolve()
            db.run(
                'insert into eventGenre (eventId, genreId) VALUES ' + placeholders,
                values,
                err => err ? reject(err) : resolve(),
            )
        })
    })
}
