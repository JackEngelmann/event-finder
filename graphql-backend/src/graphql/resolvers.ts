import { IResolvers } from "graphql-tools";
import { Database } from "sqlite3";
import { queryEvent } from "../queries/event";
import { queryEvents } from "../queries/events";
import { queryClubs } from "../queries/clubs";
import { queryClub } from "../queries/club";
import { queryGenre } from "../queries/genre";
import { queryGenres } from "../queries/genres";
import { queryGenresForEvent } from "../queries/genresForEvent";
import { createEvent } from "../mutations/createEvent";

type Source = {
    clubId: number
    id: number
}
type Context = {
    db: Database
}

export const resolvers: IResolvers<Source, Context> = {
    Query: {
        clubs: (obj, args, appContext, info) => queryClubs(appContext),
        club: (obj, args, appContext, info) => queryClub(appContext, parseInt(args.id)),
        events: (obj, args, appContext, info) => queryEvents(appContext, args.filter),
        event: (obj, args, appContext, info) => queryEvent(appContext, parseInt(args.id)),
        genre: (obj, args, appContext, info) => queryGenre(appContext, parseInt(args.id)),
        genres: (obj, args, appContext, info) => queryGenres(appContext),
    },
    Mutation: {
        createEvent: async (obj, args, appContext, info) => {
            const eventId = await createEvent(appContext, args.input)
            const event = await queryEvent(appContext, eventId)
            return { event }
        },
    },
    Event: {
        club: (event, args, appContext, info) => queryClub(appContext, event.clubId),
        genres: (event, args, appContext, info) => queryGenresForEvent(appContext, event.id)
    }
}