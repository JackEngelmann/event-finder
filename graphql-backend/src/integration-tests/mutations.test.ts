import { ApolloServer } from "apollo-server-express"
import { createTestDb } from "../database/database"
import { insertTestData, eventFragment } from "./utils"
import { createTestClient, ApolloServerTestClient } from "apollo-server-testing"
import { typeDefs } from '../graphql/schema'
import { resolvers } from '../graphql/resolvers'
import { CreateEventInput } from "../mutations/createEvent"
import { UpdateEventInput } from "../mutations/updateEvent"
import { Database } from "sqlite3"
import { queryEvent } from "../queries/event"
import { AppContext } from "../appContext"

let server: ApolloServerTestClient | undefined = undefined
let db: Database | undefined = undefined
let appContext: AppContext = { db: db! }

beforeEach(async done => {
    const db = await createTestDb()
    appContext = { db }
    await insertTestData(db)
    const newServer = new ApolloServer({
        typeDefs,
        resolvers,
        context: { db }
    })
    server = createTestClient(newServer)
    done()
})

test('create an event with only required information', async done => {
    const input: CreateEventInput = {
        name: 'event-name',
        date: 'event-date',
        clubId: 1,
    }
    const result = await server!.mutate({
        mutation: `
            mutation CreateEvent($input: CreateEventInput!) {
                createEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('create an event with empty genres', async done => {
    const input: CreateEventInput = {
        name: 'event-name',
        date: 'event-date',
        clubId: 1,
        genreIds: []
    }
    const result = await server!.mutate({
        mutation: `
            mutation CreateEvent($input: CreateEventInput!) {
                createEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('create an event with all fields', async done => {
    const input: CreateEventInput = {
        name: 'event-name',
        description: 'event-description',
        date: 'event-date',
        clubId: 1,
        genreIds: [1, 2],
        special: 'event-special',
        priceCategory: 2,
        admissionFee: 10.0,
        admissionFeeWithDiscount: 4.0,
        minimumAge: 18,
        amountOfFloors: 3
    }
    const result = await server!.mutate({
        mutation: `
            mutation CreateEvent($input: CreateEventInput!) {
                createEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('update an event to only have required fields', async done => {
    const input: UpdateEventInput = {
        id: 1,
        name: 'updated-name',
        date: 'updated-date',
        clubId: 2,
    }
    const result = await server!.mutate({
        mutation: `
            mutation UpdateEvent($input: UpdateEventInput!) {
                updateEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('return null for event when updating a non-existent event', async done => {
    const input: UpdateEventInput = {
        id: 10000,
        name: 'updated-name',
        date: 'updated-date',
        clubId: 2,
    }
    const result = await server!.mutate({
        mutation: `
            mutation UpdateEvent($input: UpdateEventInput!) {
                updateEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeNull()
    done()
})

test('update an event with empty genre list', async done => {
    const input: UpdateEventInput = {
        id: 1,
        name: 'updated-name',
        date: 'updated-date',
        clubId: 2,
        genreIds: []
    }
    const result = await server!.mutate({
        mutation: `
            mutation UpdateEvent($input: UpdateEventInput!) {
                updateEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('update all fields of an event', async done => {
    const input: UpdateEventInput = {
        id: 1,
        name: 'updated-name',
        date: 'updated-date',
        clubId: 2,
        description: 'updated-description',
        genreIds: [2, 3],
        special: 'updated-special',
        priceCategory: 1,
        admissionFee: 20.0,
        admissionFeeWithDiscount: 3.0,
        minimumAge: 16,
        amountOfFloors: 2
    }
    const result = await server!.mutate({
        mutation: `
            mutation UpdateEvent($input: UpdateEventInput!) {
                updateEvent(input: $input) {
                    event {
                        ${eventFragment}
                    }
                }
            }
        `,
        variables: { input },
    })
    expect(result.data).toBeDefined()
    expect(result).toMatchSnapshot()
    done()
})

test('delete an event', async done => {
    const id = 1

    const event = await queryEvent(appContext, id)
    expect(event).toBeDefined()

    await server!.mutate({
        mutation: `
            mutation DeleteEvent($id: Int!) {
                deleteEvent(id: $id) {
                    id
                }
            }
        `,
        variables: { id }
    })

    const deletedEvent = await queryEvent(appContext, id)
    expect(deletedEvent).toBeUndefined()
    done()
})
