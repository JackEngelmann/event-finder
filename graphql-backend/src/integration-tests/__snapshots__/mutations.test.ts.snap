// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create an event with all fields 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 4,
        "amountOfFloors": 3,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": "event-description",
        "genres": Array [
          Object {
            "id": 1,
            "name": "genre1",
          },
          Object {
            "id": 2,
            "name": "genre2",
          },
        ],
        "id": 2,
        "imageUrl": null,
        "minimumAge": 18,
        "name": "event-name",
        "priceCategory": 2,
        "special": "event-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`create an event with empty genres 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": Array [],
        "id": 2,
        "imageUrl": null,
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`create an event with only required information 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": Array [],
        "id": 2,
        "imageUrl": null,
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`create club with all information 1`] = `
Object {
  "data": Object {
    "createClub": Object {
      "club": Object {
        "address": "address",
        "contact": "contact",
        "description": "description",
        "email": "email",
        "id": 3,
        "link": "link",
        "name": "test",
        "region": "region",
        "specials": "speicals",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`create club with only required information 1`] = `
Object {
  "data": Object {
    "createClub": Object {
      "club": Object {
        "address": null,
        "contact": null,
        "description": null,
        "email": null,
        "id": 3,
        "link": null,
        "name": "test",
        "region": null,
        "specials": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`update all fields of an event 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": 20,
        "admissionFeeWithDiscount": 3,
        "amountOfFloors": 2,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": "updated-description",
        "genres": Array [
          Object {
            "id": 2,
            "name": "genre2",
          },
          Object {
            "id": 3,
            "name": "genre3",
          },
        ],
        "id": 1,
        "imageUrl": "imageUrl",
        "minimumAge": 16,
        "name": "updated-name",
        "priceCategory": 1,
        "special": "updated-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`update an event to only have required fields 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": Array [],
        "id": 1,
        "imageUrl": "imageUrl",
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`update an event with empty genre list 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": Array [],
        "id": 1,
        "imageUrl": "imageUrl",
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`update club with all information 1`] = `
Object {
  "data": Object {
    "updateClub": Object {
      "club": Object {
        "address": "updated-address",
        "contact": "updated-contact",
        "description": "updated-description",
        "email": "updated-email",
        "id": 1,
        "link": "updated-link",
        "name": "updated-test",
        "region": "updated-region",
        "specials": "updated-speicals",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`update club with only required information 1`] = `
Object {
  "data": Object {
    "updateClub": Object {
      "club": Object {
        "address": null,
        "contact": null,
        "description": null,
        "email": null,
        "id": 1,
        "link": null,
        "name": "updated-name",
        "region": null,
        "specials": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
